### ISSUE TITLE
Current temperature loaded, not temperature at training date/time

### ISSUE THREAD

**Issue by diesl** [Runalyze/Runalyze#1501]
When uploading training data from some days ago, the current temperature is loaded instead of the temperature when the training was done.

I found out that weather data is pulled from openweathermap.org, but they state that [historical data](http://openweathermap.org/history) is only available in [paid plans](http://openweathermap.org/price_detailes).

I did not dig into the code to search the API request, so I am also not sure what the reponse is like, but it would be interesting to see it. When I tried their example requests, I only got **403 Forbidden** responses.

As a consequence, I am not sure if it makes sense to insert temperature data that is wrong in most cases (if not directly after training).

**Comment by laufhannes** [comment]
Runalyze does fetch this historical data - but it's rarely available - and falls back to current data. We can drop this fallback.

```php
public function loadForecast(Location $Location) {
	$this->Result = array();

	if ($Location->isOld() && $Location->hasLocationName()) {
		$this->setFromURL( self::URL_HISTORY.'/city?q='.$Location->name().'&start='.$Location->time().'&cnt=1' );
	}

	if (empty($this->Result)) {
		if ($Location->hasPosition()) {
			$this->setFromURL( self::URL.'?lat='.$Location->lat().'&lon='.$Location->lon() );
		} elseif ($Location->hasLocationName()) {
			$this->setFromURL( self::URL.'?q='.$Location->name(), $this->cacheKey($Location->name()) );
		}
	}
}
```
(see [Openweathermap.php#L50-L64](https://github.com/Runalyze/Runalyze/blob/master/inc/core/Data/Weather/Openweathermap.php#L50-L64))

**Comment by diesl** [comment]
Thank you for the fast answer and link to the code.

I think it is misleading to fall back to current time's temperature without any notice. Sometimes it is not that obvious (large deviation or same value after multiple uploads)

> but it's rarely available

This is also what I guessed, but lately I was not sure anymore, as it is difficult to reproduce :wink:

**Comment by laufhannes** [comment]
The main problem is that Openweathermap has not enough historical data. `$Location->isOld()` checks whether the activity is more than 24 hours ago, but it's enough to run in the morning and import the activity in the afternoon to get differences of more than 10 °C these days.

We have some more issues concerning weather data open:
 - #1300 don't load weather when opening the form (date is missing, a lot of unnecessary api calls)
 - #1029 change weather when changing date

**Would it be okay to load weather data only on specific request?** 
That means: The user uploads a file or sets values by hand and next to weather's inputs there's a link "*Load weather*" which fills the inputs if an exact match is found or shows a message like "*No weather data found for this date. Current weather: 17 °C, sunny. Use this?*"

**Comment by diesl** [comment]
> We have some more issues concerning weather data open

I noticed these issues, but I didn't want to hijack them.

> Would it be okay to load weather data only on specific request?

Okay for me. The main point is that a user can not distinguish a correct historical temperature from a wrong current temperature right now.

>(We could automatically look for an exact match if "Load weather" is enabled in user's configuration and no weather is set when trying to insert the activity into our database, but do not use a fallback in that case.)

In combination with #1300 and #1029 this could also be done after date and time are set with AJAX for example.


Thinking further:
Ability to add other weather data providers (e.g. via plugins)

**Comment by laufhannes** [comment]
Just a note: historical weather data is not available anymore. Openweathermap.org requires a paid account for historical data.

**Comment by diesl** [comment]
What do you think about the idea of refactoring this as a plugin? Then the weather data provider could be changed easily, for example when there is a better service available (worldwide, or just regional)

**Comment by mipapo** [comment]
Why a Plugin? In general the weather backend could be (easily) extended by another services. Just the configuration (fallback function?) would need a refactor.

Some things shouldn't be configureable for users ;) (Used weather service) 

As the most weather apis have worldwide data it's not needed to separate between regional/worldwide. (And it's just too much for having this functionality in Runalyze. This just would make it unnecessarily complicated) 

Final thought: The configuration of weather services should get a configureable mark to see if the key has access to advanced metrics.

**Comment by mipapo** [comment]
Should be fixed in ecc1eb3f0bfc69608acf647caba9efc594fde5d5 and cf949abaffc539afab6142c2ff1ba3345e2b1305
